### PR TITLE
nose test runner improvements

### PR DESCRIPTION
**Pull request by millerdev**
Cleanup, fixes and improvements to nose test runner setup. These were made in the process of making nose to be the default test runner (not done yet). Splitting this out now because of how long that effort is taking.

Notable additions:
- make @czue's test db optimization work with nose
- a few handy nose plugins

@snopoke cc @dannyroberts

### CODE REVIEW

File: .travis/matrix-script.sh [hunk 1]
```diff
@@ -31,7 +31,9 @@ if [ "${MATRIX_TYPE}" = "python" ]; then
 
 elif [ "${MATRIX_TYPE}" = "python-sharded" ]; then
 
-    SHARDED_TEST_APPS="form_processor sql_db couchforms case phone receiverwrapper"
+    SHARDED_TEST_APPS=form_processor
+    # commented out due to bug that discarded all but first item
+    #SHARDED_TEST_APPS="form_processor sql_db couchforms case phone receiverwrapper"
```
Comment (snopoke) [inline .travis/matrix-script.sh]: I thought you fixed that bug

File: corehq/tests/noseplugins/dividedwerun.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from unittest import SkipTest
+
+from nose.case import Test, FunctionTestCase
+from nose.plugins import Plugin
+from nose.suite import ContextSuite
+
+log = logging.getLogger(__name__)
+
+
+class DividedWeRunPlugin(Plugin):
```
Comment (snopoke) [inline corehq/tests/noseplugins/dividedwerun.py]: Is the purpose of this to allow us to split the travis tests into multiple runs?

File: .travis/matrix-script.sh (hunk 1 above)
Comment (millerdev) [inline .travis/matrix-script.sh]: I didn't. The only thing I fixed was passing multiple arguments in to the test runner. Once I did that, they were failing so I changed the list to run the same tests that had been running. I thought you had worked on it, but don't remember seeing the fix come through.

File: corehq/tests/noseplugins/dividedwerun.py (hunk 2 above)
Comment (millerdev) [inline corehq/tests/noseplugins/dividedwerun.py]: yes

File: .travis/matrix-script.sh (hunk 1 above)
Comment (millerdev) [inline .travis/matrix-script.sh]: Correction: I fixed the bug in passing multiple arguments. However, I did not fix the failing tests that it uncovered.

Comment (snopoke) [inline .travis/matrix-script.sh]: https://github.com/dimagi/commcare-hq/pull/11573

Comment (millerdev) [inline .travis/matrix-script.sh]: @snopoke: curious why you took code from this PR and put it in https://github.com/dimagi/commcare-hq/pull/11573, thereby creating a merge conflict here? Are you holding off on merging this PR for some reason?

**Comment by millerdev** [comment]
good god what a disaster. the tests! failure may be related to https://github.com/dimagi/commcare-hq/pull/11573/commits/c815205acab69db365fa1e4572cd53668544dfd6#diff-d7a5c2357aa92a3498ad5a251a7be770R7?

**Comment by millerdev** [comment]
161b25e may have fixed half of the problem, and the other half is #11582

File: .travis/matrix-script.sh (hunk 1 above)
Comment (snopoke) [inline .travis/matrix-script.sh]: Applogies, should have put a 'open for review don't merge' label on that.

**Pull request merged**